code/modules/reagents/chemistry/machinery/smoke_machine.dm

```diff
@@ -33,88 +30,145 @@
-	if((!is_operational) || (!on) || (reagents.total_volume == 0))
-		icon_state = "[base_icon_state]0[panel_open ? "-o" : null]"
+	if(!is_operational || !on || !reagents.total_volume)
+		icon_state = "[base_icon_state]0[panel_open ? "-o" : ""]"
 		return ..()
+
 	icon_state = "[base_icon_state]1"
 	return ..()
 
 /obj/machinery/smoke_machine/RefreshParts()
 	. = ..()
+
+	//new capacity to store reagents from matter bins
 	var/new_volume = REAGENTS_BASE_VOLUME
 	for(var/datum/stock_part/matter_bin/matter_bin in component_parts)
 		new_volume += REAGENTS_BASE_VOLUME * matter_bin.tier
-	if(!reagents)
-		create_reagents(new_volume, INJECTABLE)
 	reagents.maximum_volume = new_volume
-	if(new_volume < reagents.total_volume)
-		reagents.expose(loc, TOUCH) // if someone manages to downgrade it without deconstructing
-		reagents.clear_reagents()
+
+	//new efficiency from capacitors
 	efficiency = 18
 	for(var/datum/stock_part/capacitor/capacitor in component_parts)
 		efficiency += 2 * capacitor.tier
+
+	//new maximum range from servos
 	max_range = 1
 	for(var/datum/stock_part/servo/servo in component_parts)
 		max_range += servo.tier
 	max_range = max(3, max_range)
 
-/obj/machinery/smoke_machine/on_set_is_operational(old_value)
-	if(old_value) //Turned off
-		end_processing()
-	else //Turned on
-		begin_processing()
+/obj/machinery/smoke_machine/item_interaction(mob/living/user, obj/item/tool, list/modifiers)
+	. = NONE
+	if(user.combat_mode || tool.item_flags & ABSTRACT || tool.flags_1 & HOLOGRAM_1 || !user.can_perform_action(src, ALLOW_SILICON_REACH))
+		return ITEM_INTERACT_SKIP_TO_ATTACK
 
+	//transfer reagents from an open container into machine
+	if(is_reagent_container(tool) && tool.is_open_container())
+		var/obj/item/reagent_containers/RC = tool
+		var/units = RC.reagents.trans_to(src, RC.amount_per_transfer_from_this, transferred_by = user)
+		if(units)
+			to_chat(user, span_notice("You transfer [units] units of the solution to [src]."))
+			return ITEM_INTERACT_SUCCESS
+		return ITEM_INTERACT_BLOCKING
 
-/obj/machinery/smoke_machine/process()
-	if(reagents.total_volume == 0)
-		on = FALSE
-		update_appearance()
+/obj/machinery/smoke_machine/wrench_act(mob/living/user, obj/item/tool)
+	. = ITEM_INTERACT_BLOCKING
+	if(on)
+		balloon_alert(user, "turn off first!")
 		return
-	var/turf/location = get_turf(src)
-	var/smoke_test = locate(/obj/effect/particle_effect/fluid/smoke) in location
-	if(on && !smoke_test)
-		update_appearance()
-		var/datum/effect_system/fluid_spread/smoke/chem/smoke_machine/smoke = new()
-		smoke.set_up(setting * 3, holder = src, location = location, carry = reagents, efficiency = efficiency)
-		smoke.start()
-		use_energy(active_power_usage)
 
-/obj/machinery/smoke_machine/wrench_act(mob/living/user, obj/item/tool)
-	. = ..()
-	if(default_unfasten_wrench(user, tool, time = 4 SECONDS))
-		on = FALSE
+	if(default_unfasten_wrench(user, tool, time = 4 SECONDS) == SUCCESSFUL_UNFASTEN)
 		return ITEM_INTERACT_SUCCESS
```

Comment:
Based on other code, it looks like it's supposed to not be possible to anchor the smoke machine while it's on, or turn it on while unanchored, but it's currently possible to circumvent this:
You just need to open the UI, start wrenching it, and then during the do_after press the on button.
This puts the smoke machine in a state where it's both on and unanchored, and impossible to re-anchor or deconstruct via screwdriver+crowbar, but thus can be dragged around while it creates smoke.

It may be needed to reintroduce code for turning the machine off after finishing unanchoring it.

Reply:
> You just need to open the UI, start wrenching it, and then during the do_after press the on button.

You cannot use/even open the UI to turn on of the machine if the machine is unanchored because the interaction flags specifically disable the UI in this state

> interaction_flags_atom = parent_type::interaction_flags_atom | INTERACT_ATOM_REQUIRES_ANCHORED

So this situation cannot occur